Adding basic middleware and web test projects.

PromontoryProtean opened the pull request:
This is a very basic ImageProcessorCore.Web project that contains middleware that does resizing of images in memory as determined by querystring parameters. It is a rough start to use as a scaffold to layer enhanced functionality onto. I also added an MVC test website that closely mirrors the test website in the current V2 repository. I did not include the file monitor code from _Layout.cshtml, however, because I wasn't sure how to replicate it in the the new aspnet core environment.

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,136 @@
+                case ".jpeg":
+                case ".jpg":
+                case ".png":
+                case ".tif":
```
There's no tif encoder/decoder

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,136 @@
+                    using (var outputStream = new MemoryStream())
+                    {
+                        //sw.Restart();
+                        image.Resize(width, height)
```
This will all need to be abstracted out to use the extension method 

    Process(this Image source, params IImageProcessor[] processors)

JimBobSquarePants commented on src/TestWebsites/MVC/Controllers/HomeController.cs:
```diff
@@ -0,0 +1,37 @@
+﻿using Microsoft.AspNet.Mvc;
```
All this should live in the tests folder in a sub directory

JimBobSquarePants commented on src/TestWebsites/MVC/Views/Home/Bmp.cshtml:
```diff
@@ -0,0 +1,99 @@
+﻿@{
+    ViewData["Title"] = "Bmp";
+}
+<article>
```
This is all using very old  [Responsive](http://responsivebp.com/) specific markup. We shouldn't need it.

Are there any new images in the diff? We don't want them.

JimBobSquarePants commented on src/TestWebsites/MVC/Views/Home/Index.cshtml:
```diff
@@ -0,0 +1,103 @@
+            <div class="col-s-6">
+                <h2>Rounded Corners</h2>
+                <img src="/images/format-Penguins.jpg?width=300&roundedcorners=26&tl=true&tr=false&bl=true&br=false" />
+            </div>
```
No rounded corners support

JimBobSquarePants commented on src/TestWebsites/MVC/Views/Home/Png.cshtml:
```diff
@@ -0,0 +1,103 @@
+            </div>
+            <div class="col-s-6">
+                <h3>comic</h3>
+                <img src="/images/format-Penguins.png?width=300&filter=comic" />
```
No comic filter

JimBobSquarePants commented on src/TestWebsites/MVC/Views/Home/Png.cshtml:
```diff
@@ -0,0 +1,103 @@
+        <div class="row">
+            <div class="col-s-6">
+                <h2>Watermark</h2>
+                <img src="/images/format-Penguins.png?width=300&watermark=text-test&color=fff&fontsize=48&fontstyle=italic&fontopacity=100&textposition=100,100&dropshadow=true&font=arial" />
```
No watermark support.

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,136 @@
+
+        public ImageProcessorMiddleware(RequestDelegate next, IApplicationEnvironment appEnvironment)
+        {
+            _next = next;
```
`this.next`

 :police_car:

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,136 @@
+
+                    //var sw = new Stopwatch();
+                    //sw.Start();
+                    var image = new Image(inputStream);
```
Resize handles correct ratio if both parameters are not passed.

JimBobSquarePants commented:
Thanks for this. It looks like there's a bit of cleanup required before I can merge the PR though.

PromontoryProtean commented:
Thanks for the feedback. I'll do some cleanup today and try to figure out how to update the pull request. I'm still getting used to GitHub.

Question on the test images. You had a format-Penguins image referenced throughout the MVC test site. I created my own images for local testing and used the same format-Penguins name so that I wouldn't have to change your markup, but should we consider including images in the repo? Or is this bad practice?

PromontoryProtean commented:
Hey Jim, I updated a bunch of files on this pull request but I am running into some problems. First I can't get `Process(this Image source, params IImageProcessor[] processors)` to work when I include resize dimensions. Second is a user error on my side, because I know you've made some commits since my pull request, and I can't figure out how to sync my fork with your changes. This will effect the naming of IImageProcessorCore back to IImageProcessor, which I see that you fixed yesterday. I'll try and get the rest of this sorted out later today or tomorrow when I have more time to debug. But I have made progress on cleaning up most of what you commented on and improved the design of the middleware quite a bit.

JimBobSquarePants commented:
Just saw your issue #321 I'll have a look at that asap. Thanks for doing some cleanup on this I'll add some comments as I go through the code.

To answer your questions:

Let's not add any additional images to the repo for now. They add bloat for anyone downloading for the first time, I'd like to actually get a nuget package going with proper test images which would test the (de/en)coders properly.

Dealing with repo changes is much easier if you use https://desktop.github.com/

Being able to simply click `Update From Core` makes stuff like the merge error go away. (Sorry about that btw, some overzealous find/replace on my part.)

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,208 @@
+            int height = 0;
+            int.TryParse(context.Request.Query["height"], out height);
+
+            var inputPath = _appEnvironment.ApplicationBasePath + "/wwwroot" + context.Request.Path;
```
I'm assuming this was just to test that you could load images? In the Framework branch we have a concept of `IImageService` that is responsible for loading images from different locations, we should copy but also improve on that approach.

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,208 @@
+            //await _next.Invoke(context);
+        }
+
+        private List<IImageProcessor> GetImageProcessorFilters(IReadableStringCollection queryString)
```
While this works it's not extensible. In Framework a collection of `IWebGraphicsProcessor` are loaded via the `BuildManager` which allows for plugins. That behaviour should be copied and improved on.

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,208 @@
+            if (!isInt)
+                return false;
+
+            if (number > 100 || number < 0)
```
SHould be handled within `Alpha`

JimBobSquarePants commented on src/ImageProcessorCore.Web/Middleware/ImageProcessorMiddleware.cs:
```diff
@@ -0,0 +1,208 @@
+            return processors;
+        }
+
+        public bool IsImageProcessorRequest(HttpContext context)
```
Some image requests can be extensionless.

JimBobSquarePants commented:
I think it's best to strip this right back so we have a raw Middleware instance only and a single page with one image. We can discuss/design the process from there.

PromontoryProtean commented:
I had not planned on adding any extra processors. I ended up doing that when I could not get `image.Process` to work. I basically used that extra code to help create https://github.com/JimBobSquarePants/ImageProcessor/issues/321. It really didn't belong in the pull request, but it got synced because I am still learning Git and a bit sloppy.

When you say raw middleware instance with only a single page, are you thinking we strip the middleware back to a simple operation like I did originally with resize? And then adjust the MVC pages to only perform that single operation? My goal from the beginning was a simple start to open the discussion about APIs, so I definitely agree with you. I think that would make the most sense, and kind of goes with the idea that I had in the original disclaimer in the readme that I included. You mentioned IWebGraphicsProcess, and I have been wondering about that myself. No sense for me to charge forward with anything until I completely understand the architecture and API plans that you have. 

So let me know if I interpreted your post about stripping things back correctly and I'll make some more changes. Thanks for the suggestions about the desktop app, I'll check that out.

PromontoryProtean commented:
I never heard back on my last post here, but I had some time this morning so I stepped things back quite a bit. Hopefully I'm on the right track. Can't really get any more basic than what is in place now.

JimBobSquarePants commented:
Cripes! Sorry, I didn't even see a notification. I'll have a look at it all today.

JimBobSquarePants commented on tests/TestWebsites/MVC/Controllers/HomeController.cs:
```diff
@@ -0,0 +1,32 @@
+            return View();
+        }
+
+        public IActionResult Bmp()
```
We only have the one view now no?

JimBobSquarePants commented on tests/TestWebsites/MVC/Views/Shared/_Layout.cshtml:
```diff
@@ -0,0 +1,26 @@
+<head>
+    <meta charset="utf-8" />
+    <title>@ViewBag.Title Processing Samples</title>
+    <link href="~/css/responsive.min.css" rel="stylesheet" type="text/css" />
```
Let's drop responsive here and clean up the views. Then we'll be good to go.

PromontoryProtean commented:
Ok, made some more changes to clean everything up. No more responsive CSS, I simplified the views and removed some dependencies that didn't need to be there in project.json and Startup.js.

What do you think would be the smartest thing to tackle next? It seems to me that we need to figure out if we are going to include a new implementation of IWebGraphicsProcessor. Or do you want to hold off on the web stuff until you make some more progress on the core? I know you are knee deep in the jpeg stuff right now.

JimBobSquarePants commented:
Excellent. We should be able to merge this now and start fixing API issues in core to ensure we can actually use it on a server. 

I think you're right, Let's get an `IWebImageProcessor` (We'll call it that to match the core library) interface designed properly, figure out if things like `BuildManager` still exist so we can load extensions and get a really sensible url API designed.

One of the joys of being able to design the API is that we can simply drop in the new jpeg encoder once I crack it so other work can still carry on.

PromontoryProtean commented:
Glad to hear we are close on this. Let me know if you find any more snags or want to make additional changes before you can merge it.

I checked out the GitHub desktop app you recommended and managed to screw up my local repository. It detected different file changes than VS. Sorry for the off topic question, but do you use the desktop app exclusively and ignore the VS integration? I'm getting the feeling that if I try to use both the desktop app and VS git integration that they might conflict with one another.

Anyway, I'm thinking once we get the initial commit done we can start collaborating on the API design. What is the best way to discuss? Start a new pull request and go from there?

JimBobSquarePants commented:
We'll raise an issue and mark it as a discussion first before a PR I think.

I never use the VS integration, just the desktop app. I find if I mix at all I always end up with issues.

JimBobSquarePants commented:
Hey there, hope all is well. 

Would it be posible for you to repoint this at my new Core-Web branch? I think I have an idea of how to proceed.

PromontoryProtean commented:
Ok, will do. I've been tied up today but I should have a chance in the next day or so.

JimBobSquarePants commented:
Excellent Thanks!

JimBobSquarePants commented:
I'll close this off now I have merged the other.
